chunky/src/java/se/llbit/math/PackedOctree.java

```diff
@@ -246,41 +294,55 @@ private boolean nodeEquals(int firstNodeIndex, Octree.Node secondNode) {
     return ((firstIsBranch && secondIsBranch) || -treeData[firstNodeIndex] == secondNode.type); // compare types (don't forget that in the tree the negation of the type is stored)
   }
 
+  /**
+   * Sets a specified block within the octree to a specific palette value, subdividing and merging as needed.
+   *
+   * x, y, z are in octree coordinates, NOT world coordinates.
+   */
   @Override
   public void set(int type, int x, int y, int z) {
     set(new Octree.Node(type), x, y, z);
   }
 
+  /**
+   * Sets a specified block within the octree to a specific palette value, subdividing and merging as needed.
+   *
+   * x, y, z are in octree coordinates, NOT world coordinates.
+   */
   @Override
   public void set(Octree.Node data, int x, int y, int z) {
-    int[] parents = new int[depth]; // better to put as a field to preventallocation at each invocation?
-    int nodeIndex = 0;
-    int parentLevel = depth - 1;
-    int position = 0;
+    int[] parents = new int[depth]; // better to put as a field to prevent allocation at each invocation?
+    int nodeIndex = 0; // start at root
+    int position;
+
+    // root is at the end of the array, its direct parent is at the front.
     for(int i = depth - 1; i >= 0; --i) {
       parents[i] = nodeIndex;
 
-      if(nodeEquals(nodeIndex, data)) {
+      if(nodeEquals(nodeIndex, data)) { // Everything in this region is already of this blocktype.
         return;
-      } else if(treeData[nodeIndex] <= 0) { // It's a leaf node
+      }
+
+      if(treeData[nodeIndex] <= 0) { // It's a leaf node
         subdivideNode(nodeIndex);
-        parentLevel = i;
       }
 
+      // Determine index of child (to go to next)
       int xbit = 1 & (x >> i);
       int ybit = 1 & (y >> i);
       int zbit = 1 & (z >> i);
       position = (xbit << 2) | (ybit << 1) | zbit;
       nodeIndex = treeData[nodeIndex] + position;
 
     }
-    int finalNodeIndex = treeData[parents[0]] + position;
-    treeData[finalNodeIndex] = -data.type; // Store negation of the type
+    // store type into final node (this specific block coordinate's node)
+    treeData[nodeIndex] = -data.type; // Negation of BlockPalette type stored
 
-    // Merge nodes where all children have been set to the same type.
-    for(int i = 0; i <= parentLevel; ++i) {
+    // Merge nodes where all children have been set to the same type, starting from the bottom.
+    for(int i = 0; i < depth; ++i) {
```

Comment:
Note: This is equivalent and less obscure; loop will break anyways if goes too high.